common/src/voxel_math.rs

```diff
@@ -89,3 +107,154 @@ impl IndexMut<CoordAxis> for Coords {
         self.0.index_mut(coord_axis as usize)
     }
 }
+
+/// Represents one of the six main directions within a chunk: positive or negative x, y, and z.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ChunkDirection {
+    pub axis: CoordAxis,
+    pub sign: CoordSign,
+}
+
+impl ChunkDirection {
+    pub const PLUS_X: Self = ChunkDirection {
+        axis: CoordAxis::X,
+        sign: CoordSign::Plus,
+    };
+    pub const PLUS_Y: Self = ChunkDirection {
+        axis: CoordAxis::Y,
+        sign: CoordSign::Plus,
+    };
+    pub const PLUS_Z: Self = ChunkDirection {
+        axis: CoordAxis::Z,
+        sign: CoordSign::Plus,
+    };
+    pub const MINUS_X: Self = ChunkDirection {
+        axis: CoordAxis::X,
+        sign: CoordSign::Minus,
+    };
+    pub const MINUS_Y: Self = ChunkDirection {
+        axis: CoordAxis::Y,
+        sign: CoordSign::Minus,
+    };
+    pub const MINUS_Z: Self = ChunkDirection {
+        axis: CoordAxis::Z,
+        sign: CoordSign::Minus,
+    };
+
+    pub fn iter() -> impl ExactSizeIterator<Item = ChunkDirection> {
+        [
+            Self::PLUS_X,
+            Self::PLUS_Y,
+            Self::PLUS_Z,
+            Self::MINUS_X,
+            Self::MINUS_Y,
+            Self::MINUS_Z,
+        ]
+        .iter()
+        .copied()
+    }
+}
+
+/// Represents one of the 48 possible orientations a chunk can be viewed from, including reflections.
+/// This is analogous to a 3x3 rotation matrix with a restricted domain.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ChunkOrientation {
+    directions: [ChunkDirection; 3],
+}
+
+impl Index<CoordAxis> for ChunkOrientation {
+    type Output = ChunkDirection;
+
+    fn index(&self, index: CoordAxis) -> &Self::Output {
+        &self.directions[index as usize]
+    }
+}
+
+impl std::ops::Mul<ChunkDirection> for ChunkOrientation {
+    type Output = ChunkDirection;
+
+    fn mul(self, rhs: ChunkDirection) -> Self::Output {
+        ChunkDirection {
+            axis: self[rhs.axis].axis,
+            sign: self[rhs.axis].sign * rhs.sign,
+        }
+    }
+}
+
+impl std::ops::Mul<ChunkOrientation> for ChunkOrientation {
+    type Output = ChunkOrientation;
+
+    fn mul(self, rhs: ChunkOrientation) -> Self::Output {
+        ChunkOrientation {
+            directions: rhs.directions.map(|d| self * d),
+        }
+    }
+}
+
+impl std::ops::MulAssign for ChunkOrientation {
+    fn mul_assign(&mut self, rhs: Self) {
+        *self = *self * rhs;
+    }
+}
+
+/// Represents one of the 6 possible orientations a chunk can be viewed from, including reflections, that preserve the origin.
+/// In other words, any instance of this struct represents a permutation of the axes.
+/// This is analogous to a 3x3 rotation matrix with a restricted domain.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SimpleChunkOrientation {
```

Comment:
This name is kind of opaque. When do we want to use it vs. the non-simple kind?

Reply:
I'm rather unsure about the name as well. It could be called `ChunkAxisPermutation` or `OriginPreservingChunkOrientation`.

Now that I got rid of `ChunkOrientation`, this could take its place, but I'm not convinced `ChunkOrientation` will be gone forever, so that seems risky.

If I don't think of anything better, I'll probably rename this to `ChunkAxisPermutation`.